File: android-core/src/androidTest/java/com/mparticle/UploadMessageTest.java
package com.mparticle;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import android.os.Handler;
import android.os.Looper;

import com.mparticle.internal.Constants;
import com.mparticle.internal.MPUtility;
import com.mparticle.internal.MParticleApiClientImpl;
import com.mparticle.networking.Matcher;
import com.mparticle.networking.MockServer;
import com.mparticle.networking.Request;
import com.mparticle.testutils.BaseCleanStartedEachTest;
import com.mparticle.testutils.MPLatch;
import com.mparticle.testutils.TestingUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;


public final class UploadMessageTest extends BaseCleanStartedEachTest {

    /**
     * set MPID, log between 0 and 20 random MPEvents, and check to make sure each one is properly
     * attributed to the correct MPID, and there are no duplicates
     */
    @Test
    public void testCorrectlyAttributeEventsToMpid() throws Exception {
        int numberOfEvents = 3;
        final Handler handler = new Handler(Looper.getMainLooper());
        long mpid = ran.nextLong();
        MParticle.getInstance().Internal().getConfigManager().setMpid(mpid, ran.nextBoolean());
        final Map<String, MPEvent> events = new HashMap<String, MPEvent>();
        final CountDownLatch latch = new MPLatch(numberOfEvents);

        final Map<Long, Map<String, JSONObject>> matchingJSONEvents = new HashMap<Long, Map<String, JSONObject>>();
        com.mparticle.internal.AccessUtils.setMParticleApiClient(new com.mparticle.internal.AccessUtils.EmptyMParticleApiClient() {
            @Override
            public int sendMessageBatch(final String message) throws IOException, MParticleApiClientImpl.MPThrottleException, MParticleApiClientImpl.MPRampException {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            JSONObject jsonObject = new JSONObject(message);
                            JSONArray jsonArray = jsonObject.optJSONArray(Constants.MessageKey.MESSAGES);

                            long mpid = Long.valueOf(jsonObject.getString("mpid"));
                            Map<String, JSONObject> matchingMpidJSONEvents = matchingJSONEvents.get(mpid);
                            if (matchingMpidJSONEvents == null) {
                                matchingJSONEvents.put(mpid, matchingMpidJSONEvents = new HashMap<String, JSONObject>());
                            }
                            if (!MPUtility.isEmpty(jsonArray)) {
                                for (int i = 0; i < jsonArray.length(); i++) {
                                    JSONObject eventObject = jsonArray.getJSONObject(i);
                                    if (eventObject.getString("dt").equals(Constants.MessageType.EVENT)) {
                                        String eventName = eventObject.getString("n");
                                        MPEvent matchingEvent = events.get(eventName);
                                        if (matchingEvent != null) {
                                            String eventType = eventObject.getString("et");
                                            if (matchingEvent.getEventType().toString().equals(eventType)) {
                                                if (matchingMpidJSONEvents.containsKey(eventName)) {
                                                    fail("Duplicate Event Message Sent");
                                                } else {
                                                    matchingMpidJSONEvents.put(eventName, eventObject);
                                                }
                                            } else {
                                                fail("Unknown Event");
                                            }
                                        } else {
                                            fail("Unknown Event");
                                        }
                                        latch.countDown();
                                    }
                                }
                            }

                        } catch (Exception e) {
                            e.printStackTrace();
                            fail(e.toString());
                        }
                    }
                });
                return 202;
            }
        });

        for (int j = 0; j < 3; j++) {
            MPEvent event = TestingUtils.getInstance().getRandomMPEventRich();
            if (events.containsKey(event.getEventName())) {
                j--;
            } else {
                events.put(event.getEventName(), event);
                MParticle.getInstance().logEvent(event);
            }
        }

        MParticle.getInstance().upload();
        latch.await();

        Map<String, JSONObject> jsonMap = matchingJSONEvents.get(mpid);
        if (events.size() > 0) {
            assertNotNull(jsonMap);
        }
        if (events != null && events.size() != 0 && events.size() != jsonMap.size()) {
            assertEquals(events.size(), jsonMap.size());
        }
    }

    @Test
    public void testEventAccuracy() throws Exception {
        final Handler handler = new Handler(Looper.getMainLooper());
        final Map<String, MPEvent> receivedEvents = new HashMap<String, MPEvent>();
        final Map<String, JSONObject> sentEvents = new HashMap<String, JSONObject>();
        final CountDownLatch latch = new MPLatch(1);
        mServer.waitForVerify(new Matcher(mServer.Endpoints().getEventsUrl()), new MockServer.RequestReceivedCallback() {
            @Override
            public void onRequestReceived(Request request) {
                try {
                    JSONObject jsonObject = request.getBodyJson();
                    JSONArray jsonArray = jsonObject.optJSONArray(Constants.MessageKey.MESSAGES);
                    if (!MPUtility.isEmpty(jsonArray)) {
                        for (int i = 0; i < jsonArray.length(); i++) {
                            JSONObject eventObject = jsonArray.getJSONObject(i);
                            if (eventObject.getString("dt").equals(Constants.MessageType.EVENT)) {
                                String eventName = eventObject.getString("n");
                                if (sentEvents.containsKey(eventName)) {
                                    fail("Duplicate Event");
                                } else {
                                    sentEvents.put(eventName, eventObject);
                                }
                            }
                        }
                    }

                } catch (Exception e) {
                    e.printStackTrace();
                    fail(e.toString());
                }
                if (sentEvents.size() == receivedEvents.size())
                latch.countDown();
            }
        });

        for (int j = 0; j < 3; j++) {
            MPEvent event = TestingUtils.getInstance().getRandomMPEventRich();
            if (receivedEvents.containsKey(event.getEventName())) {
                j--;
            } else {
                receivedEvents.put(event.getEventName(), event);
                MParticle.getInstance().logEvent(event);
            }
        }

        MParticle.getInstance().upload();
        latch.await();
        for (Map.Entry<String, MPEvent> entry : receivedEvents.entrySet()) {
            if (!sentEvents.containsKey(entry.getKey())) {
                assertNull(entry.getValue());
            } else {
                assertTrue(sentEvents.containsKey(entry.getKey()));
                JSONObject jsonObject = sentEvents.get(entry.getKey());
                assertEventEquals(entry.getValue(), jsonObject);
            }
        }
    }

    void assertEventEquals(MPEvent mpEvent, JSONObject jsonObject) throws JSONException {
        if (jsonObject.optString("n") != mpEvent.getEventName()) {
            assertTrue(mpEvent.getEventName().equals(jsonObject.getString("n")));
        }
        if (mpEvent.getLength() != null || jsonObject.has("el")) {
            assertEquals(mpEvent.getLength(), jsonObject.getDouble("el"), .1);
        }
        if (!mpEvent.getEventType().toString().equals(jsonObject.optString("et"))) {
            assertTrue(mpEvent.getEventType().toString().equals(jsonObject.getString("et")));
        }

        Map<String, String> customAttributesTarget = mpEvent.getCustomAttributeStrings() == null ? new HashMap<String, String>() : mpEvent.getCustomAttributeStrings();
        JSONObject customAttributes = jsonObject.optJSONObject("attrs");
        if (customAttributes != null) {
            Iterator<String> keysIterator = customAttributes.keys();
            while (customAttributes != null && keysIterator.hasNext()) {
                String key = customAttributes.keys().next();
                String jsonVal = keysIterator.next();
                String objVal = customAttributesTarget.get(key);
                if (jsonVal != objVal) {
                    String val = customAttributes.getString(key);
                    if (!val.equals(customAttributesTarget.get(key)) && !key.equals("EventLength") && !(val.equals("null") && objVal == null)) {
                        assertTrue(customAttributes.getString(key).equals(customAttributesTarget.get(key)));
                    }
                }
            }
        }

        Map<String, List<String>> customFlagTarget = mpEvent.getCustomFlags();
        JSONObject customFlags = jsonObject.optJSONObject("flags");
        if (customFlags != null) {
            Iterator<String> flagsIterator = customFlags.keys();
            while (flagsIterator.hasNext()) {
                String key = flagsIterator.next();
                JSONArray values = customFlags.getJSONArray(key);
                List<String> flags = customFlagTarget.get(key);
                assertArraysEqual(values, flags);
            }
        }
    }


    void assertArraysEqual(JSONArray jsonArray, List<String> list) throws JSONException {
        List<String> jsonArrayList = new ArrayList<String>();
        for (int i = 0; i < jsonArray.length(); i++) {
            jsonArrayList.add(jsonArray.getString(i));
        }
        assertEquals(list.size(), jsonArrayList.size());
        Collections.sort(list);
        Collections.sort(jsonArrayList);
        for (int i = 0; i < list.size(); i++) {
            String a = list.get(i);
            String b = jsonArrayList.get(i);
            if (a == null) {
                assertTrue(b.equals("null"));
            } else {
                assertTrue(a.equals(b));
            }
        }
    }
}

File: android-core/src/androidTest/java/com/mparticle/UploadMessageTest.kt
package com.mparticle

import android.os.Handler
import android.os.Looper
import com.mparticle.internal.Constants
import com.mparticle.internal.EmptyApiClient
import com.mparticle.internal.MPUtility
import com.mparticle.messages.events.MPEventMessage
import com.mparticle.testing.BaseStartedTest
import com.mparticle.testing.FailureLatch
import com.mparticle.testing.mockserver.EndpointType
import com.mparticle.testing.mockserver.Server
import com.mparticle.testing.orThrow
import com.mparticle.utils.randomMPEventRich
import com.mparticle.utils.setMParticleApiClient
import org.json.JSONException
import org.json.JSONObject
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import java.util.Collections
import java.util.concurrent.CountDownLatch
import kotlin.random.Random

class UploadMessageTest : BaseStartedTest() {
    /**
     * set MPID, log between 0 and 20 random MPEvents, and check to make sure each one is properly
     * attributed to the correct MPID, and there are no duplicates
     */
    @Test
    @Throws(Exception::class)
    fun testCorrectlyAttributeEventsToMpid() {
        val numberOfEvents = 3
        val handler: Handler = Handler(Looper.getMainLooper())
        val mpid: Long = Random.Default.nextLong()
        MParticle.getInstance().orThrow().Internal().configManager.setMpid(mpid, Random.Default.nextBoolean())
        val events: MutableMap<String, MPEvent> = HashMap()
        val latch: CountDownLatch = FailureLatch(count = numberOfEvents)
        val matchingJSONEvents: MutableMap<Long, MutableMap<String, JSONObject>> = HashMap()
        setMParticleApiClient(object : EmptyApiClient() {
            override fun sendMessageBatch(message: String): Int {
                handler.post {
                    try {
                        val jsonObject = JSONObject(message)
                        val jsonArray = jsonObject.optJSONArray(Constants.MessageKey.MESSAGES)
                        val mpid = java.lang.Long.valueOf(jsonObject.getString("mpid"))
                        var matchingMpidJSONEvents = matchingJSONEvents[mpid]
                        if (matchingMpidJSONEvents == null) {
                            matchingJSONEvents[mpid] =
                                HashMap<String, JSONObject>().also { matchingMpidJSONEvents = it }
                        }
                        if (!MPUtility.isEmpty(jsonArray)) {
                            for (i in 0 until jsonArray.length()) {
                                val eventObject = jsonArray.getJSONObject(i)
                                if (eventObject.getString("dt") == Constants.MessageType.EVENT) {
                                    val eventName = eventObject.getString("n")
                                    val matchingEvent = events[eventName]
                                    if (matchingEvent != null) {
                                        val eventType = eventObject.getString("et")
                                        if (matchingEvent.eventType.toString() == eventType) {
                                            if (matchingMpidJSONEvents!!.containsKey(eventName)) {
                                                Assert.fail("Duplicate Event Message Sent")
                                            } else {
                                                matchingMpidJSONEvents!![eventName] = eventObject
                                            }
                                        } else {
                                            Assert.fail("Unknown Event")
                                        }
                                    } else {
                                        Assert.fail("Unknown Event")
                                    }
                                    latch.countDown()
                                }
                            }
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                        Assert.fail(e.toString())
                    }
                }
                return 202
            }
        })
        var j = 0
        while (j < 3) {
            val event: MPEvent = randomMPEventRich()
            if (events.containsKey(event.eventName)) {
                j--
            } else {
                events[event.eventName] = event
                MParticle.getInstance()!!.logEvent(event)
            }
            j++
        }
        MParticle.getInstance()!!.upload()
        latch.await()
        val jsonMap: Map<String, JSONObject> = matchingJSONEvents[mpid]!!
        if (events.size > 0) {
            Assert.assertNotNull(jsonMap)
        }
        if (events != null && events.size != 0 && events.size != jsonMap.size) {
            assertEquals(events.size.toLong(), jsonMap.size.toLong())
        }
    }

    @Test
    @Throws(Exception::class)
    fun testEventAccuracy() {
        val receivedEvents = mutableMapOf<String, MPEventMessage>()
        val sentEvents: MutableMap<String, MPEvent> = HashMap()
        Server
            .endpoint(EndpointType.Events)
            .assertWillReceive {
                it.body.messages
                    .filterIsInstance<MPEventMessage>()
                    .let {
                        it.forEach { event ->
                            assertNull(receivedEvents[event.name])
                            receivedEvents[event.name!!] = event
                        }
                    }
                receivedEvents.size == sentEvents.size
            }
            .after {
                (0..3).forEach {
                    val event: MPEvent = randomMPEventRich()
                    if (!sentEvents.containsKey(event.eventName)) {
                        sentEvents[event.eventName] = event
                        MParticle.getInstance()!!.logEvent(event)
                    }
                }
                MParticle.getInstance()?.upload()
            }
            .blockUntilFinished()

        sentEvents.values.forEach {
            assertEventEquals(it, receivedEvents[it.eventName])
        }
    }

    @Throws(JSONException::class)
    fun assertEventEquals(mpEvent: MPEvent, eventMessage: MPEventMessage?) {
        if (eventMessage == null) {
            assertNotNull(eventMessage)
            return
        }
        if (eventMessage.name !== mpEvent.eventName) {
            Assert.assertTrue(mpEvent.eventName == eventMessage.name)
        }
        if (mpEvent.length != null || eventMessage.eventDuration != null) {
            assertEquals(mpEvent.length!!, eventMessage.eventDuration!!, .1)
        }
        if (mpEvent.eventType.name != eventMessage.eventType?.name) {
            assertEquals(mpEvent.eventType.name, eventMessage.eventType?.name)
        }
        val customAttributesTarget =
            if (mpEvent.customAttributes == null) mapOf() else mpEvent.customAttributes!!
        val customAttributes = mpEvent.customAttributes ?: mapOf()
        assertEquals(customAttributesTarget.size, customAttributes.size)
        customAttributes
            .filter { it.value != customAttributesTarget[it.key] }
            .filter { it.value == "null" && customAttributesTarget[it.key] == null }
            .filter { it.key != "EventLength" }
            .let {
                assertEquals("extra keys in message: ${it.entries.joinToString { it.toString() }}", 0, it.size)
            }
        customAttributesTarget
            .filter { !customAttributes.containsKey(it.key) }.let {
                assertEquals("missing keys in message: ${it.entries.joinToString { it.toString() }}", 0, it.size)
            }
        val customFlagTarget = mpEvent.customFlags ?: mapOf()
        val customFlags = eventMessage.eventFlags
        if (customFlags != null) {
            customFlags.keys
                .forEach {
                    with(customFlags[it]) {
                        when (this) {
                            is List<*> -> assertArraysEqual(this, customFlagTarget[it] ?: listOf())
                            else -> assertEquals(this.toString(), customFlagTarget[it].toString())
                        }
                    }
                }
        }
    }

    @Throws(JSONException::class)
    fun assertArraysEqual(messageValue: List<*>, testValue: List<String>) {
        val jsonArrayList: MutableList<String> = ArrayList()
        for (i in 0 until messageValue.size) {
            jsonArrayList.add(messageValue.get(i).toString())
        }
        assertEquals(testValue.size.toLong(), jsonArrayList.size.toLong())
        Collections.sort(testValue)
        Collections.sort(jsonArrayList)
        for (i in testValue.indices) {
            val a = testValue[i]
            val b = messageValue[i]
            if (a == null) {
                assertEquals(b, "null")
            } else {
                assertEquals(a, b)
            }
        }
    }
}
